src/blockdata/witness.rs

```diff
@@ -0,0 +1,327 @@
+//! Witness
+//!
+//! This module contains the [`Witness`] struct and relative methods to operate on it
+//!
+
+use consensus::encode::{Error, MAX_VEC_SIZE};
+use consensus::{Decodable, Encodable, WriteExt};
+use io::{self, Read, Write};
+use prelude::*;
+use VarInt;
+
+#[cfg(feature = "serde")] use serde;
+
+/// The Witness is the data used to unlock bitcoins since the [segwit upgrade](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki)
+///
+/// Can be logically seen as an array of byte-arrays `Vec<Vec<u8>>` and indeed you can convert from
+/// it and collect the iteration to convert into it.
+/// For serialization and deserialization performance it is stored internally as a single `Vec`,
+/// saving some allocations
+///
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub struct Witness {
+    /// contains the witness Vec<Vec<u8>> serialization without the initial varint indicating the
+    /// number of elements (which is stored in len)
+    content: Vec<u8>,
+
+    /// Number of elements in the witness.
+    /// It is stored separately (instead of as VarInt in the initial part of content) so that method
+    /// like [`Witness::push`] doesn't have case requiring to shift the entire array
+    witness_elements: u64,
+}
+
+/// Support structure to allow efficient and convenient iteration over the Witness elements
+pub struct WitnessIterator<'a> {
+    witness: &'a Witness,
+    cursor: usize,
```

Comment:
Why not just `core::slice::Iter<'a, u8>`?

Reply:
You can, it has [`as_slice()`](https://doc.rust-lang.org/std/slice/struct.Iter.html#method.as_slice) method. I used it in #662 and #673

Was also thinking of doing [`take_slice_or_kill`](https://github.com/rust-bitcoin/rust-bitcoin/pull/673/files#diff-af6864f932193bfbb57ebf73cd8c41bc12a33a25e3b27fa47ce95c62392158daR649) via extension trait, so it could be reused. You seem to be doing something similar so maybe worth revisiting that idea.

(Was also considering adding it to `core` as it seems widely applicable.)